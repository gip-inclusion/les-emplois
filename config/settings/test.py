import os


ITOU_ENVIRONMENT = "DEV"
os.environ["ITOU_ENVIRONMENT"] = ITOU_ENVIRONMENT

from .base import *  # noqa: E402,F403


SECRET_KEY = "foobar"
ALLOWED_HOSTS = []

# We *want* to do the same `collectstatic` on the CI than on PROD to catch errors early,
# but we don't want to do it when running the test suite locally for performance reasons.
if not os.getenv("CI", False):
    # `ManifestStaticFilesStorage` (used in base settings) requires `collectstatic` to be run.
    STORAGES["staticfiles"]["BACKEND"] = "django.contrib.staticfiles.storage.StaticFilesStorage"  # noqa: F405

ASP_ITOU_PREFIX = "XXXXX"  # same as in our fixtures
ITOU_PROTOCOL = "http"
ITOU_FQDN = "localhost:8000"

DATABASES["default"]["HOST"] = os.getenv("PGHOST", "127.0.0.1")  # noqa: F405
DATABASES["default"]["PORT"] = os.getenv("PGPORT", "5432")  # noqa: F405
DATABASES["default"]["NAME"] = os.getenv("PGDATABASE", "itou")  # noqa: F405
DATABASES["default"]["USER"] = os.getenv("PGUSER", "postgres")  # noqa: F405
DATABASES["default"]["PASSWORD"] = os.getenv("PGPASSWORD", "password")  # noqa: F405

MAILJET_API_KEY_PRINCIPAL = "API_MAILJET_KEY_PRINCIPAL"
MAILJET_SECRET_KEY_PRINCIPAL = "API_MAILJET_SECRET_PRINCIPAL"

AWS_S3_ENDPOINT_URL = f"http://{os.getenv('CELLAR_ADDON_HOST', 'localhost:9000')}/"
AWS_S3_ACCESS_KEY_ID = "minioadmin"
AWS_S3_SECRET_ACCESS_KEY = "minioadmin"
AWS_STORAGE_BUCKET_NAME = "tests"
