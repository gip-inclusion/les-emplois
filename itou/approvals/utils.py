from itou.companies.enums import SIAE_WITH_CONVENTION_KINDS


def get_user_last_accepted_siae_job_application(user):
    if not user.is_job_seeker:
        return None

    # Some candidates may not have accepted job applications
    # Assuming it's the case can lead to issues downstream
    return (
        user.job_applications.accepted()
        .filter(to_company__kind__in=SIAE_WITH_CONVENTION_KINDS)
        .with_accepted_at()
        .order_by("-accepted_at", "-hiring_start_at")
        .first()
    )


def last_hire_was_made_by_siae(user, siae):
    if not user.is_job_seeker:
        return False
    last_accepted_job_application = get_user_last_accepted_siae_job_application(user)
    return last_accepted_job_application and last_accepted_job_application.to_company_id == siae.pk
