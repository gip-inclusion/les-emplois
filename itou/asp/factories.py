import datetime
import random

import factory

from itou.asp import models


_sample_europe_countries = [
    {"code": "101", "name": "DANEMARK", "group": "2"},
    {"code": "111", "name": "BULGARIE", "group": "2"},
    {"code": "135", "name": "PAYS-BAS", "group": "2"},
]

_sample_outside_europe_countries = [
    {"code": "212", "name": "AFGHANISTAN", "group": "3"},
    {"code": "436", "name": "BAHAMAS", "group": "3"},
    {"code": "324", "name": "CONGO", "group": "3"},
]

_sample_france = [
    {"code": "714", "name": "BORA-BORA", "group": "1"},
    {"code": "812", "name": "KOUMAC", "group": "1"},
    {"code": "737", "name": "PUKAPUKA", "group": "1"},
]

_sample_communes = [
    {"code": "64483", "name": "SAINT-JEAN-DE-LUZ"},
    {"code": "97108", "name": "CAPESTERRE-DE-MARIE-GALANT"},
    {"code": "97107", "name": "CAPESTERRE-BELLE-EAU"},
    {"code": "37273", "name": "VILLE-AUX-DAMES"},
    {"code": "13200", "name": "MARSEILLE"},
    {"code": "67152", "name": "GEISPOLSHEIM"},
    {"code": "85146", "name": "MONTAIGU"},
]


class CountryFactory(factory.django.DjangoModelFactory):
    class Meta:
        model = models.Country
        django_get_or_create = ("code",)

    code, name, group = random.choice(_sample_europe_countries).values()


class CountryFranceFactory(CountryFactory):
    code, name, group = "100", "FRANCE", "1"


class CountryCollectiviteOutremerFactory(CountryFactory):
    """
    Parts or regions of France having their own INSEE country code

    Most of them are "Collectivit√©s d'Outre-Mer"
    """

    code, name, group = random.choice(_sample_france).values()


class CountryEuropeFactory(CountryFactory):
    code, name, group = random.choice(_sample_europe_countries).values()


class CountryOutsideEuropeFactory(CountryFactory):
    code, name, group = random.choice(_sample_outside_europe_countries).values()


class CommuneFactory(factory.django.DjangoModelFactory):
    """Factory for ASP INSEE commune:
    - if `code` or `name` are build parameters, object values will result of a lookup in `_sample_communes`
    - otherwise, fields `code` and `name` will be set from a randomly picked sample commune
    """

    class Meta:
        model = models.Commune
        django_get_or_create = (
            "code",
            "start_date",
            "end_date",
        )

    # FIXME: may cause issues in testing validity periods
    start_date = datetime.date(2000, 1, 1)
    end_date = None

    # Skipping definition of `code` and `name` fields
    # will be set in `_adjust_kwargs`

    @classmethod
    def _adjust_kwargs(cls, **kwargs):
        # Allow creation with parameters `code` or `name` (first matched)
        # either field must be a match in `_sample_communes`
        match kwargs:
            case {"code": code}:
                for item in _sample_communes:
                    if item["code"] == code:
                        kwargs["name"] = item["name"]
            case {"name": name}:
                for item in _sample_communes:
                    if item["name"] == name:
                        kwargs["code"] = item["code"]
            case _:
                code, name = random.choice(_sample_communes).values()
                kwargs["code"] = code
                kwargs["name"] = name

        return kwargs

    @classmethod
    def _adjust_kwargs(cls, **kwargs):
        # Allow creation with parameters `code` or `name` (first matched)
        # either field must be a match in `_sample_communes`
        match kwargs:
            case {"code": code}:
                for item in _sample_communes:
                    if item["code"] == code:
                        kwargs["name"] = item["name"]
            case {"name": name}:
                for item in _sample_communes:
                    if item["name"] == name:
                        kwargs["code"] = item["code"]
            case _:
                code, name = random.choice(_sample_communes).values()
                kwargs["code"] = code
                kwargs["name"] = name

        return kwargs


# FIXME: unreliable and confusing
class MockedCommuneFactory(CommuneFactory):
    """
    A factory with a specific code for mock testing
    """

    code, name = _sample_communes[-1].values()
