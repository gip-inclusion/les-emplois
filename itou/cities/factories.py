import json
import os

from django.contrib.gis.geos import GEOSGeometry, Point
from django.template.defaultfilters import slugify

from itou.cities.models import City


CURRENT_DIR = os.path.dirname(os.path.realpath(__file__))

CITIES_JSON_FILE = f"{CURRENT_DIR}/management/commands/data/cities.json"


def create_test_cities(selected_departments, num_per_department=None):
    """
    Not a factory strictly speaking, but it's sitting here for discoverability.

    Pick cities from `cities.json` based on the given department numbers.
        create_test_cities(['01', '02'])

    The number of cities in each department can be limited:
        create_test_cities(['54', '57'], num_per_department=2)
    """
    departments_counter = {department: 0 for department in selected_departments}

    with open(CITIES_JSON_FILE, "r") as raw_json_data:

        json_data = json.load(raw_json_data)

        for item in json_data:

            department = item.get("codeDepartement")
            if (not department) or (department not in selected_departments):
                continue

            coords = item.get("centre")
            if not coords:
                continue

            departments_counter[department] += 1
            if num_per_department and departments_counter[department] > num_per_department:
                if all(value >= num_per_department for value in departments_counter.values()):
                    break
                continue

            City.objects.create(
                slug=slugify(f"{item['nom']}-{department}"),
                department=department,
                name=item["nom"],
                post_codes=item["codesPostaux"],
                code_insee=item["code"],
                coords=GEOSGeometry(f"{coords}"),  # Feed `GEOSGeometry` with GeoJSON.
            )


def create_city_saint_andre():
    return City.objects.create(
        name="Saint-André-des-Eaux",
        slug="saint-andre-des-eaux-44",
        department="44",
        coords=Point(-2.3140436, 47.3618584),
        post_codes=["44117"],
        code_insee="44117",
    )


def create_city_guerande():
    return City.objects.create(
        name="Guérande",
        slug="guerande-44",
        department="44",
        coords=Point(-2.4747713, 47.3358576),
        # Dummy
        post_codes=["44350"],
        code_insee="44350",
    )


def create_city_vannes():
    return City.objects.create(
        name="Vannes",
        slug="vannes-56",
        department="56",
        coords=Point(-2.8186843, 47.657641),
        # Dummy
        post_codes=["56000"],
        code_insee="56000",
    )
