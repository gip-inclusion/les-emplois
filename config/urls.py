from django.conf import settings
from django.contrib import admin
from django.urls import include, path, re_path, register_converter
from django.views.generic import TemplateView

from itou.utils.urls import SiretConverter
from itou.www.dashboard import views as dashboard_views
from itou.www.login import views as login_views
from itou.www.signup import views as signup_views


register_converter(SiretConverter, "siret")

urlpatterns = [
    path("admin/", admin.site.urls),
    # --------------------------------------------------------------------------------------
    # allauth URLs. Order is important because some URLs are overriden.
    # --------------------------------------------------------------------------------------
    # Override allauth `account_signup` URL.
    # /accounts/signup/ <=> account_signup
    # We don't want any user to be able to signup using the default allauth `signup` url
    # because we have multiple specific signup processes for different kind of users.
    re_path(r"^accounts/signup/$", signup_views.signup),
    # --------------------------------------------------------------------------------------
    # Override allauth `account_login` URL.
    # /accounts/login/ <=> account_login
    # We override this view because the login page should look slightly differently
    # for job seekers, prescribers and employers.
    # Also, PEAMU is only available for job seekers.
    re_path(r"^accounts/login/$", login_views.login),
    # --------------------------------------------------------------------------------------
    # Override allauth `account_change_password` URL.
    # /accounts/password/change/ <=> account_change_password
    # https://github.com/pennersr/django-allauth/issues/468
    re_path(r"^accounts/password/change/$", dashboard_views.password_change),
    # --------------------------------------------------------------------------------------
    # Override allauth `account_reset_password` URL.
    # Avoid user enumeration via password reset page.
    re_path(r"^accounts/password/reset/$", signup_views.ItouPasswordResetView.as_view()),
    # --------------------------------------------------------------------------------------
    # Override allauth `account_logout` URL.
    # /accounts/logout/ <=> account_logout
    # We need custom code to process PEAMU logout.
    re_path(r"^accounts/logout/$", dashboard_views.logout),
    # --------------------------------------------------------------------------------------
    # Other allauth URLs.
    path("accounts/", include("allauth.urls")),
    # --------------------------------------------------------------------------------------
    # PEAMU URLs.
    path("accounts/", include("itou.allauth_adapters.peamu.urls")),
    # FranceConnect URLs.
    path("franceconnect/", include("itou.france_connect.urls")),
    # --------------------------------------------------------------------------------------
    # API.
    path("api/v1/", include("itou.api.urls", namespace="v1")),
    path("api/", include("itou.www.api.urls")),
    # www.
    path("", include("itou.www.home.urls")),
    path("apply/", include("itou.www.apply.urls")),
    path("approvals/", include("itou.www.approvals_views.urls")),
    path("autocomplete/", include("itou.www.autocomplete.urls")),
    path("dashboard/", include("itou.www.dashboard.urls")),
    path("employee_record/", include("itou.www.employee_record_views.urls")),
    path("institutions/", include("itou.www.institutions_views.urls")),
    path("invitations/", include("itou.www.invitations_views.urls")),
    path("prescribers/", include("itou.www.prescribers_views.urls")),
    path("search/", include("itou.www.search.urls")),
    path("siae/", include("itou.www.siaes_views.urls")),
    path("signup/", include("itou.www.signup.urls")),
    path("stats/", include("itou.www.stats.urls")),
    path("versions/", include("itou.www.releases.urls")),
    # Enable Anymailâ€™s status tracking
    # https://anymail.readthedocs.io/en/stable/esps/mailjet/#status-tracking-webhooks
    re_path(r"^webhooks/anymail/", include("anymail.urls")),
    path("welcoming_tour/", include("itou.www.welcoming_tour.urls")),
    # Static pages.
    path("accessibility/", TemplateView.as_view(template_name="static/accessibility.html"), name="accessibility"),
    # Errors pages.
    path("403/", TemplateView.as_view(template_name="403.html"), name="403"),
    path("404/", TemplateView.as_view(template_name="404.html"), name="404"),
    path("500/", TemplateView.as_view(template_name="500.html"), name="500"),
]

if settings.DEBUG and "debug_toolbar" in settings.INSTALLED_APPS:
    import debug_toolbar

    urlpatterns = [path("__debug__/", include(debug_toolbar.urls))] + urlpatterns
    # In the local developpemnt environment, we enable the default callback URLs
    urlpatterns = [path("", include("itou.france_connect.urls"))] + urlpatterns
