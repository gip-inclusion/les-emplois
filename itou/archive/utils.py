import datetime

from django.db import models
from django.db.models import Count, Exists, OuterRef, Q
from django.db.models.functions import Coalesce
from django.utils import timezone

from itou.approvals.models import Approval
from itou.eligibility.models import EligibilityDiagnosis, GEIQEligibilityDiagnosis
from itou.gps.models import FollowUpGroup
from itou.job_applications.models import JobApplication
from itou.users.models import User, UserKind


def get_filter_kwargs_on_user_for_related_objects_to_check():
    """
    Returns a dictionary of filter parameters to check for objects related
    to the User model that are not cascade deleted.
    """
    return {
        f"{obj.name}__isnull": True
        for obj in User._meta.related_objects
        if getattr(obj, "on_delete", None) and obj.on_delete != models.CASCADE
    }


def get_year_month_or_none(date=None):
    if not date:
        return None

    if isinstance(date, datetime.datetime):
        return timezone.localdate(date).replace(day=1)

    return date.replace(day=1)


def count_related_subquery(model, fk_field, outer_ref_field, extra_filters=None):
    filters = {fk_field: OuterRef(outer_ref_field)}
    if extra_filters:
        filters.update(extra_filters)
    return Coalesce(
        model.objects.filter(**filters).values(fk_field).annotate(count=Count(outer_ref_field)).values("count"), 0
    )


def inactive_jobseekers_without_recent_related_objects(inactive_since, notified, batch_size=None):
    recent_approval = Approval.objects.filter(
        Q(end_at__gt=inactive_since) | Q(updated_at__gt=inactive_since), user_id=OuterRef("pk")
    )
    recent_eligibility_diagnosis = EligibilityDiagnosis.objects.filter(
        Q(expires_at__gt=inactive_since) | Q(updated_at__gt=inactive_since), job_seeker_id=OuterRef("pk")
    )
    recent_geiq_eligibility_diagnosis = GEIQEligibilityDiagnosis.objects.filter(
        Q(expires_at__gt=inactive_since) | Q(updated_at__gt=inactive_since), job_seeker_id=OuterRef("pk")
    )
    recent_job_application = JobApplication.objects.filter(job_seeker_id=OuterRef("pk"), updated_at__gt=inactive_since)
    recent_followup_group = FollowUpGroup.objects.filter(beneficiary_id=OuterRef("pk"), updated_at__gt=inactive_since)

    qs = (
        User.objects.filter(
            (Q(last_login__lte=inactive_since) | Q(last_login__isnull=True)),
            date_joined__lte=inactive_since,
            kind=UserKind.JOB_SEEKER,
            upcoming_deletion_notified_at__isnull=not notified,
        )
        .filter(
            ~Exists(recent_approval),
            ~Exists(recent_eligibility_diagnosis),
            ~Exists(recent_geiq_eligibility_diagnosis),
            ~Exists(recent_job_application),
            ~Exists(recent_followup_group),
        )
        .order_by("date_joined", "pk")
    )

    if batch_size:
        return qs[:batch_size]

    return qs
