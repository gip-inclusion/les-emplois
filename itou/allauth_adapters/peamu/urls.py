from django.urls import include, re_path

from itou.allauth_adapters.peamu import views as peamu_views
from itou.allauth_adapters.peamu.provider import PEAMUProvider


def default_urlpatterns(provider):
    urlpatterns = [
        re_path(r"^login/$", peamu_views.oauth2_login, name=provider.id + "_login"),
        re_path(r"^login/callback/$", peamu_views.oauth2_callback, name=provider.id + "_callback"),
    ]

    return [
        re_path(r"^profile/$", peamu_views.redirect_to_dashboard_view),
        re_path(r"^" + provider.get_slug() + r"/", include(urlpatterns)),
    ]


urlpatterns = default_urlpatterns(PEAMUProvider)
