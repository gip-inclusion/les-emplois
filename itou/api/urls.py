from django.urls import include, path
from drf_spectacular.views import SpectacularAPIView, SpectacularRedocView
from rest_framework import routers
from rest_framework.authtoken import views as auth_views

from itou.api.data_inclusion_api.views import DataInclusionStructureView

from .applicants_api.views import ApplicantsView
from .employee_record_api.viewsets import (
    DummyEmployeeRecordViewSet,
    EmployeeRecordUpdateNotificationViewSet,
    EmployeeRecordViewSet,
)
from .siae_api.viewsets import SiaeViewSet


# High level app for API
# https://docs.djangoproject.com/en/dev/topics/http/urls/#url-namespaces-and-included-urlconfs
app_name = "itou.api"

# Using DRF router with viewsets means automatic definition of utl patterns
router = routers.DefaultRouter()
router.register(r"employee-records", EmployeeRecordViewSet, basename="employee-records")
router.register(
    r"employee-record-notifications",
    EmployeeRecordUpdateNotificationViewSet,
    basename="employee-record-notifications",
)
router.register(r"dummy-employee-records", DummyEmployeeRecordViewSet, basename="dummy-employee-records")
router.register(r"siaes", SiaeViewSet, basename="siaes")

urlpatterns = [
    # TokenAuthentication endpoint to get token from login/password.
    path("token-auth/", auth_views.obtain_auth_token, name="token-auth"),
    # Needed for Browseable API (dev)
    path("api-auth/", include("rest_framework.urls", namespace="rest_framework")),
    # OpenAPI
    # See: https://www.django-rest-framework.org/topics/documenting-your-api/
    # OAS 3 YAML schema (downloadable)
    path(
        "oas3/",
        SpectacularAPIView.as_view(),
        name="openapi_schema",
    ),
    path(
        "redoc/",
        SpectacularRedocView.as_view(url_name="v1:openapi_schema"),
        name="redoc",
    ),
]

urlpatterns += router.urls

urlpatterns += [
    path("structures/", DataInclusionStructureView.as_view(), name="structures-list"),
    path("candidats/", ApplicantsView.as_view(), name="applicants-list"),
]
