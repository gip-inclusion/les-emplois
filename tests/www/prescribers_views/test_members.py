from django.core import mail
from django.urls import reverse

from tests.prescribers.factories import (
    PrescriberFactory,
    PrescriberMembershipFactory,
    PrescriberOrganizationFactory,
    PrescriberOrganizationWith2MembershipFactory,
    PrescriberOrganizationWithMembershipFactory,
)
from tests.utils.test import TestCase


class MembersTest(TestCase):
    def test_members(self):
        organization = PrescriberOrganizationWithMembershipFactory()
        user = organization.members.first()
        self.client.force_login(user)
        url = reverse("prescribers_views:members")
        response = self.client.get(url)
        assert response.status_code == 200

    def test_active_members(self):
        organization = PrescriberOrganizationFactory()
        active_member_active_user = PrescriberMembershipFactory(organization=organization)
        active_member_inactive_user = PrescriberMembershipFactory(organization=organization, user__is_active=False)
        inactive_member_active_user = PrescriberMembershipFactory(organization=organization, is_active=False)
        inactive_member_inactive_user = PrescriberMembershipFactory(
            organization=organization, is_active=False, user__is_active=False
        )

        self.client.force_login(active_member_active_user.user)
        url = reverse("prescribers_views:members")
        response = self.client.get(url)
        assert response.status_code == 200
        assert len(response.context["members"]) == 1
        assert active_member_active_user in response.context["members"]
        assert active_member_inactive_user not in response.context["members"]
        assert inactive_member_active_user not in response.context["members"]
        assert inactive_member_inactive_user not in response.context["members"]


class UserMembershipDeactivationTest(TestCase):
    def test_self_deactivation(self):
        """
        A user, even if admin, can't self-deactivate
        (must be done by another admin)
        """
        organization = PrescriberOrganizationWithMembershipFactory()
        admin = organization.members.filter(prescribermembership__is_admin=True).first()
        memberships = admin.prescribermembership_set.all()
        membership = memberships.first()

        self.client.force_login(admin)
        url = reverse("prescribers_views:deactivate_member", kwargs={"user_id": admin.id})
        response = self.client.post(url)
        assert response.status_code == 403

        # Trying to change self membership is not allowed
        # but does not raise an error (does nothing)
        membership.refresh_from_db()
        assert membership.is_active

    def test_deactivate_user(self):
        """
        Standard use case of user deactivation.
        Everything should be fine ...
        """
        organization = PrescriberOrganizationWith2MembershipFactory()
        admin = organization.members.filter(prescribermembership__is_admin=True).first()
        guest = organization.members.filter(prescribermembership__is_admin=False).first()

        memberships = guest.prescribermembership_set.all()
        membership = memberships.first()

        self.client.force_login(admin)
        url = reverse("prescribers_views:deactivate_member", kwargs={"user_id": guest.id})
        response = self.client.post(url)
        assert response.status_code == 302

        # User should be deactivated now
        membership.refresh_from_db()
        assert not membership.is_active
        assert admin == membership.updated_by
        assert membership.updated_at is not None

        # Check mailbox
        # User must have been notified of deactivation (we're human after all)
        assert len(mail.outbox) == 1
        email = mail.outbox[0]
        assert f"[Désactivation] Vous n'êtes plus membre de {organization.display_name}" == email.subject
        assert "Un administrateur vous a retiré d'une structure sur les emplois de l'inclusion" in email.body
        assert email.to[0] == guest.email

    def test_deactivate_with_no_perms(self):
        """
        Non-admin user can't change memberships
        """
        organization = PrescriberOrganizationWithMembershipFactory()
        guest = PrescriberFactory()
        organization.members.add(guest)

        self.client.force_login(guest)
        url = reverse("prescribers_views:deactivate_member", kwargs={"user_id": guest.id})
        response = self.client.post(url)
        assert response.status_code == 403

    def test_deactivated_prescriber_is_orienter(self):
        """
        A prescriber deactivated from a prescriber organization
        and without any membership becomes an "orienteur".
        As such he must be able to login.
        """
        organization = PrescriberOrganizationWith2MembershipFactory()
        admin = organization.members.filter(prescribermembership__is_admin=True).first()
        guest = organization.members.filter(prescribermembership__is_admin=False).first()

        self.client.force_login(admin)
        url = reverse("prescribers_views:deactivate_member", kwargs={"user_id": guest.id})
        response = self.client.post(url)
        assert response.status_code == 302

        # guest is now an orienter
        self.client.force_login(guest)
        url = reverse("dashboard:index")
        response = self.client.post(url)
        assert response.status_code == 200

    def test_structure_selector(self):
        """
        Check that a deactivated member can't access the structure
        from dashboard selector
        """
        organization2 = PrescriberOrganizationWithMembershipFactory()
        guest = organization2.members.first()

        organization1 = PrescriberOrganizationWithMembershipFactory()
        admin = organization1.members.first()
        organization1.members.add(guest)

        memberships = guest.prescribermembership_set.all()
        assert len(memberships) == 2

        # Admin remove guest from structure
        self.client.force_login(admin)
        url = reverse("prescribers_views:deactivate_member", kwargs={"user_id": guest.id})
        response = self.client.post(url)
        assert response.status_code == 302
        self.client.logout()

        # guest must be able to login
        self.client.force_login(guest)
        url = reverse("dashboard:index")
        response = self.client.get(url)

        # Wherever guest lands should give a 200 OK
        assert response.status_code == 200

        # Check response context, only one prescriber organization should remain
        assert len(response.context["request"].organizations) == 1
