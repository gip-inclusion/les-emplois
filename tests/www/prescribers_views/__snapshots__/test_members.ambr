# serializer version: 1
# name: TestMembers.test_members
  '''
  <main class="s-main" id="main" role="main">
      <div aria-atomic="true" aria-live="polite" class="toast-container">
      </div>
      <section class="s-title-02">
          <div class="s-title-02__container container">
              <div class="s-title-02__row row">
                  <div class="s-title-02__col col-12">
                      <div class="c-title">
                          <div class="c-title__main">
                              <h1>
                                  Organisation
                              </h1>
                          </div>
                      </div>
                  </div>
              </div>
          </div>
      </section>
      <section class="s-section">
          <div class="s-section__container container">
              <div class="s-section__row row">
                  <div class="s-section__col col-12">
                      <div class="d-flex flex-column flex-lg-row gap-3 align-items-lg-center justify-content-lg-between mb-3">
                          <h2 class="mb-0">
                              Collaborateurs
                          </h2>
                          <div aria-label="Actions sur les collaborateurs" class="d-flex flex-column flex-md-row gap-2 justify-content-md-end" role="group">
                              <a class="btn btn-primary btn-ico" href="/invitations/invite_prescriber_with_org">
                                  <i aria-hidden="true" class="ri-user-add-line ri-lg">
                                  </i>
                                  <span>
                                      Ajouter des collaborateurs
                                  </span>
                              </a>
                          </div>
                      </div>
                      <p>
                          Vous êtes connecté(e) en tant que
                          <b>
                              Pierre DUPONT
                          </b>
                          (pierre.dupont@test.local)
                      </p>
                      <p class="mb-0">
                          1 collaborateur (dont 1 administrateur)
                      </p>
                      <div class="table-responsive mt-3 mt-md-4">
                          <table class="table table-hover">
                              <caption class="visually-hidden">
                                  Liste des collaborateurs
                              </caption>
                              <thead>
                                  <tr>
                                      <th scope="col">
                                          Nom
                                      </th>
                                      <th scope="col">
                                          Email
                                      </th>
                                      <th scope="col">
                                          Date d'inscription
                                      </th>
                                      <th class="text-end w-50px" scope="col">
                                      </th>
                                  </tr>
                              </thead>
                              <tbody>
                                  <tr>
                                      <td>
                                          Pierre DUPONT
                                          <button class="badge badge-xs rounded-pill bg-info" data-bs-title="Administrateur de la structure" data-bs-toggle="tooltip" type="button">
                                              Admin
                                              <span class="visually-hidden">
                                                  istrateur de la structure
                                              </span>
                                          </button>
                                      </td>
                                      <td>
                                          <a class="btn-link" href="mailto:pierre.dupont@test.local">
                                              pierre.dupont@test.local
                                          </a>
                                      </td>
                                      <td>
                                          12 juillet 2025 à 12:40
                                      </td>
                                      <td class="text-end w-50px">
                                          <button class="btn btn-sm btn-link btn-ico-only disabled" data-bs-title="Aucune action possible" data-bs-toggle="tooltip" type="button">
                                              <i aria-label="Aucune action possible" class="ri-more-2-fill">
                                              </i>
                                          </button>
                                      </td>
                                  </tr>
                              </tbody>
                          </table>
                      </div>
                  </div>
              </div>
          </div>
      </section>
  </main>
  
  '''
# ---
# name: TestUserMembershipDeactivation.test_deactivate_user
  '''
  Un administrateur vous a retiré d'une structure sur les emplois de l'inclusion
  
  Organisation :
  
  - Nom : Mission locale
  - Type : FT
  
  - Email de contact : ml@mailinator.com
  
  Si vous estimez qu'il peut s'agir d'une erreur, contactez un des administrateurs cette structure sur les emplois de l'inclusion.
  
  ---
  [DEV] Cet email est envoyé depuis un environnement de démonstration, merci de ne pas en tenir compte [DEV]
  Les emplois de l'inclusion
  http://localhost:8000/
  '''
# ---
